routing/pathfind.go

```diff
@@ -573,10 +573,24 @@ func findPath(g *graphParams, r *RestrictParams, cfg *PathFindingConfig,
 	// if the cltv limit is MaxUint32.
 	absoluteCltvLimit := uint64(r.CltvLimit) + uint64(finalHtlcExpiry)
 
-	// Calculate the absolute attempt cost that is used for probability
-	// estimation.
-	absoluteAttemptCost := int64(cfg.AttemptCost) +
-		int64(amt)*cfg.AttemptCostPPM/1000000
+	// Calculate the default attempt cost as configured globally.
+	defaultAttemptCost := float64(
+		cfg.AttemptCost +
+			amt*lnwire.MilliSatoshi(cfg.AttemptCostPPM)/1000000,
+	)
+
+	// Validate time preference value.
+	if math.Abs(timePref) > 1 {
+		return nil, fmt.Errorf("time preference %v out of range [-1, 1]",
+			timePref)
+	}
+
+	// Scale to avoid the extremes -1 and 1 which run into infinity issues.
+	timePref *= 0.99
+
+	// Apply time preference. At 0, the default attempt cost will
+	// be used.
+	absoluteAttemptCost := defaultAttemptCost * (1/(0.5-timePref/2) - 1)
```

Comment:
This is where it happens.

Reply:
Question is whether we want to change it, or just set the goal of this pr to introducing the `time_pref` parameter and refine its behavior later on?

Reply:
I would probably set it to a safer lower value in the beginning, perhaps 0.9 is already enough (max factor of 19).
If we compare a hop with P=0.5 with one of P=1, we get 1/0.5 - 1/1 = 1 as a difference. With a default base_attempt and and rate_attempt of 100 msat and 1000PPM, this results in a difference in 1000.1 sat for an amount of 1000000 sat, i.e., you pay that amount more in fees for knowing that the P=1 hop succeeds with respect to the P=0.5 hop. A factor of 19 is more than enough to exaggerate usage of the more probable channel, I guess. Hope my calculations are correct there.

Reply:
Fine with me to change to 0.9. The reason to be cautious is that you expect users to use 1 as a value and that they'll be surprised by the resulting route even when a fee limit is set?

Reply:
Updated to 0.9

Reply:
Very cool how this works out in the end: a multiplier that scales the attempt cost, with a larger value meaning the attempt cost is increased which'll scale proportionally with the estimated probability.